x-pack/plugins/enterprise_search/public/applications/shared/telemetry/telemetry_logic.test.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { resetContext } from 'kea';
+
+import { JSON_HEADER as headers } from '../../../../common/constants';
+import { mockHttpValues } from '../../__mocks__';
+jest.mock('../http', () => ({
+  HttpLogic: { values: { http: mockHttpValues.http } },
+}));
+
+import { TelemetryLogic } from './';
+
+describe('Telemetry logic', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+    resetContext({});
+    TelemetryLogic.mount();
+  });
+
+  describe('sendTelemetry', () => {
+    it('successfully calls the server-side telemetry endpoint', () => {
+      TelemetryLogic.actions.sendTelemetry({
+        action: 'viewed',
+        metric: 'setup_guide',
+        product: 'enterprise_search',
+      });
+
+      expect(mockHttpValues.http.put).toHaveBeenCalledWith('/api/enterprise_search/stats', {
+        headers,
+        body: '{"product":"enterprise_search","action":"viewed","metric":"setup_guide"}',
+      });
+    });
+
+    it('throws an error if the telemetry endpoint fails', () => {
+      mockHttpValues.http.put.mockImplementationOnce(() => Promise.reject());
+      try {
+        TelemetryLogic.actions.sendTelemetry({ action: '', metric: '', product: '' });
+      } catch (e) {
+        expect(e.message).toEqual('Unable to send telemetry');
```

Comment:
Just a check, could this test pass with a false positive in the case where this action never threw?

Like if someone updated the telemetry method to use `post` instead of `put`, I think this would still pass... because `put` is never called, the action never throws, and this catch is never executed. In that case, this test should fail but it would pass.

Just thinking that something written like the following might be more resilient:

```ts
    it('throws an error if the telemetry endpoint fails', async () => {
      const promise = Promise.reject()
      mockHttpValues.http.put.get.mockReturnValue(promise);
      let message = "";
      TelemetryLogic.actions.sendTelemetry({ action: '', metric: '', product: '' });

      try {
        await promise;
      } catch (e) {
        message = e.message;
      }
      expect(message).toEqual('Unable to send telemetry');
    });
```

Actually, I'm wondering if this is already a false positive. As in, since this is not an async test method, is that catch ever even being triggered?

Reply:
Ohh super good point about false positives/this not being an async fn! 👍 I like your proposed solution a lot, will definitely implement it!

Reply:
You were super on point about it being a false positive! I added an `expect.assertions(1)` to the top of the test and confirmed that it had 0 assertions, which means the catch wasn't firing. Credit to this article for the nifty technique:

https://medium.com/@liran.tal/demystifying-jest-async-testing-patterns-b730d4cca4ec#a5c3